### ISSUE TITLE
Coverity Issues

### ISSUE THREAD

**Issue by jdinan** [Sandia-OpenSHMEM/SOS#768]
Coverity had quite a lot to say about the recent changes to the benchmarks:
```
New defect(s) Reported-by: Coverity Scan
Showing 20 of 28 defect(s)


** CID 188152:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/latency_common.h: 104 in multi_size_latency()


________________________________________________________________________________________________________
*** CID 188152:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/latency_common.h: 104 in multi_size_latency()
98     
99     /*have symmetric buffers src/dest from perf_metrics_t
100      *  that has been initialized to my_node number */
101     extern void streaming_latency(int len, perf_metrics_t *data);
102     
103     static inline 
>>>     CID 188152:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter data of type "perf_metrics_t" (size 152 bytes) by value.
104     void multi_size_latency(perf_metrics_t data, char *argv[]) {
105         unsigned int len;
106         int partner_pe = partner_node(data);
107     
108         if (data.my_node == 0) {
109             print_latency_header();

** CID 188151:    (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/shmem_latency_get_perf.c: 60 in int_element_latency()
/test/performance/shmem_perf_suite/shmem_latency_put_perf.c: 60 in int_element_latency()


________________________________________________________________________________________________________
*** CID 188151:    (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/shmem_latency_get_perf.c: 60 in int_element_latency()
54     #ifndef USE_NONBLOCKING_API
55         long_element_round_trip_latency_get(data);
56     #endif
57     }
58     
59     void
>>>     CID 188151:    (PASS_BY_VALUE)
>>>     Passing parameter data of type "perf_metrics_t" (size 152 bytes) by value.
60     int_element_latency(perf_metrics_t data)
61     {
62     #ifndef USE_NONBLOCKING_API
63         int_g_latency(data);
64     #endif
65     }
/test/performance/shmem_perf_suite/shmem_latency_put_perf.c: 60 in int_element_latency()
54     #ifndef USE_NONBLOCKING_API
55         long_element_round_trip_latency_put(data);
56     #endif
57     }
58     
59     void
>>>     CID 188151:    (PASS_BY_VALUE)
>>>     Passing parameter data of type "perf_metrics_t" (size 152 bytes) by value.
60     int_element_latency(perf_metrics_t data)
61     {
62     #ifndef USE_NONBLOCKING_API
63         int_p_latency(data);
64     #endif
65     }

** CID 188150:    (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/latency_common.h: 53 in calc_and_print_results()
/test/performance/shmem_perf_suite/bw_common.h: 144 in calc_and_print_results()


________________________________________________________________________________________________________
*** CID 188150:    (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/latency_common.h: 53 in calc_and_print_results()
47         printf("%4sin bytes%17sin us\n", " ", " ");
48     }
49     
50     /* calculation and printing of the latency */
51     static inline 
52     void calc_and_print_results(double start, double end, int len,
>>>     CID 188150:    (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
53                                 perf_metrics_t metric_info) {
54         int stride = 0, start_pe = 0, nPEs = 0;
55         int nred_elements = 1;
56         static double latency = 0.0, avg_latency = 0.0;
57         static double pwrk[SHMEM_REDUCE_MIN_WRKDATA_SIZE];
58         
/test/performance/shmem_perf_suite/bw_common.h: 144 in calc_and_print_results()
138     
139         printf("\n");
140     }
141     
142     static inline 
143     void calc_and_print_results(double end_t, double start_t, int len, 
>>>     CID 188150:    (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
144                                 perf_metrics_t metric_info) {
145         int stride = 0, start_pe = 0, nPEs = 0;
146         static double pe_bw_sum, bw = 0.0; /*must be symmetric for reduction*/
147         double pe_bw_avg = 0.0, pe_mr_avg = 0.0;
148         int nred_elements = 1;
149         static double pwrk[SHMEM_REDUCE_MIN_WRKDATA_SIZE];

** CID 188149:  Integer handling issues  (OVERFLOW_BEFORE_WIDEN)
/test/apps/gups.c: 427 in SHMEMRandomAccess()


________________________________________________________________________________________________________
*** CID 188149:  Integer handling issues  (OVERFLOW_BEFORE_WIDEN)
/test/apps/gups.c: 427 in SHMEMRandomAccess()
421       /* Default number of global updates to table: 4x number of table entries */
422       if (NumUpdatesOpt == 0) {
423          ProcNumUpdates = 4 * LocalTableSize;
424          NumUpdates = 4 * TableSize;
425       } else {
426          ProcNumUpdates = NumUpdatesOpt;
>>>     CID 188149:  Integer handling issues  (OVERFLOW_BEFORE_WIDEN)
>>>     Potentially overflowing expression "NumUpdatesOpt * NumProcs" with type "int" (32 bits, signed) is evaluated using 32-bit arithmetic, and then used in a context that expects an expression of type "uint64_t" (64 bits, unsigned).
427          NumUpdates = NumUpdatesOpt * NumProcs;
428       }
429     
430       if (MyProc == 0) {
431         fprintf( outFile, "Running on %d processors\n", NumProcs);
432         fprintf( outFile, "Total Main table size = 2^%" PRIu64 " = %" PRIu64 " words\n",

** CID 188148:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/bw_common.h: 90 in print_bw_header()


________________________________________________________________________________________________________
*** CID 188148:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/bw_common.h: 90 in print_bw_header()
84         }
85     
86         printf("%15s in Mops/sec%15s  in us\n", " ", " ");
87     }
88     
89     static 
>>>     CID 188148:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
90     void print_bw_header(perf_metrics_t metric_info) {
91         print_header(metric_info);
92         printf("\n\nBandwidth test type:    %10s\n", metric_info.bw_type_str);
93     
94         printf("\nMessage Size%15sBandwidth%15sMessage Rate\n", " ", " ");
95     

** CID 188147:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/bw_common.h: 331 in uni_dir_bw_test_and_output()


________________________________________________________________________________________________________
*** CID 188147:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/bw_common.h: 331 in uni_dir_bw_test_and_output()
325     /* have one symmetric char array metric_info->buf of max_len to use for
326      * calculation initalized with my_node number
327      * NOTE: post function validation assumptions, data isn't flushed pre/post */
328     extern void uni_dir_bw(int len, perf_metrics_t *metric_info);
329     
330     static inline 
>>>     CID 188147:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
331     void uni_dir_bw_test_and_output(perf_metrics_t metric_info) {
332         int partner_pe = partner_node(metric_info);
333         unsigned long int len = 0;
334     
335         if(metric_info.my_node == 0) {
336             if (metric_info.opstyle == STYLE_ATOMIC)

** CID 188146:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/bw_common.h: 288 in bi_dir_bw_test_and_output()


________________________________________________________________________________________________________
*** CID 188146:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/bw_common.h: 288 in bi_dir_bw_test_and_output()
282     /*have two symmetric char array metric_info->src/dest of max_len to
283      * use for calculation initalized with my_node number
284      * NOTE: post function validation assumptions, data isn't flushed pre/post */
285     extern void bi_dir_bw(int len, perf_metrics_t *metric_info);
286     
287     static inline 
>>>     CID 188146:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
288     void bi_dir_bw_test_and_output(perf_metrics_t metric_info) {
289         int partner_pe = partner_node(metric_info);
290         unsigned long int len;
291     
292         if(metric_info.my_node == 0) {
293             if (metric_info.opstyle == STYLE_ATOMIC)

** CID 188145:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/round_t_latency.h: 29 in long_element_round_trip_latency_get()


________________________________________________________________________________________________________
*** CID 188145:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/round_t_latency.h: 29 in long_element_round_trip_latency_get()
23      * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
24      * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
25      * SOFTWARE.
26      */
27     
28     static inline
>>>     CID 188145:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
29     void long_element_round_trip_latency_get(perf_metrics_t metric_info)
30     {
31         double start = 0.0;
32         double end = 0.0;
33         int dest = partner_node(metric_info);
34         int receiver = (metric_info.num_pes != 1) ? streaming_node(metric_info) : true;

** CID 188144:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/round_t_latency.h: 78 in long_element_round_trip_latency_put()


________________________________________________________________________________________________________
*** CID 188144:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/round_t_latency.h: 78 in long_element_round_trip_latency_put()
72                             *metric_info.target, dest);
73             }
74         }
75     } /*gauge small get pathway round trip latency*/
76     
77     static inline
>>>     CID 188144:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
78     void long_element_round_trip_latency_put(perf_metrics_t metric_info)
79     {
80         double start = 0.0;
81         double end = 0.0;
82         long tmp;
83         int dest = partner_node(metric_info);

** CID 188143:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/common.h: 613 in check_hostname_validation()


________________________________________________________________________________________________________
*** CID 188143:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/common.h: 613 in check_hostname_validation()
607             assert(my_info.cstyle == COMM_INCAST);
608             return true;
609         }
610     }
611     
612     static inline
>>>     CID 188143:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter my_info of type "perf_metrics_t" (size 152 bytes) by value.
613     int check_hostname_validation(perf_metrics_t my_info) {
614     
615         int hostname_status = -1;
616     
617         /* hostname_size should be a length divisible by 4 */
618         int hostname_size = (MAX_HOSTNAME_LEN % 4 == 0) ? MAX_HOSTNAME_LEN :

** CID 188142:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/bw_common.h: 109 in print_data_results()


________________________________________________________________________________________________________
*** CID 188142:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/bw_common.h: 109 in print_data_results()
103         }
104     
105         printf("%16sin msgs/sec\n", " ");
106     }
107     
108     static 
>>>     CID 188142:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter data of type "perf_metrics_t" (size 152 bytes) by value.
109     void print_data_results(double bw, double mr, perf_metrics_t data,
110                                 int len, double total_t) {
111         static int atomic_type_index = 0;
112     
113         if (data.opstyle == STYLE_ATOMIC) {
114             printf("%-10s", dt_names[atomic_type_index]);

** CID 188141:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/common.h: 766 in PE_set_used_adjustments()


________________________________________________________________________________________________________
*** CID 188141:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/common.h: 766 in PE_set_used_adjustments()
760     }
761     
762     /* reduction to collect performance results from PE set
763      * then start_pe will print results --- assumes num_pes is even */
764     static inline
765     void PE_set_used_adjustments(int *nPEs, int *stride, int *start_pe,
>>>     CID 188141:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter my_info of type "perf_metrics_t" (size 152 bytes) by value.
766                                  perf_metrics_t my_info) {
767         red_PE_set PE_set = validation_set(my_info, nPEs);
768     
769         if(PE_set == FIRST_HALF || PE_set == FULL_SET) {
770             *start_pe = 0;
771         }

** CID 188140:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/common.h: 741 in validation_set()


________________________________________________________________________________________________________
*** CID 188140:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/common.h: 741 in validation_set()
735             metric_info->warmup = WARMUP_LARGE;
736         }
737     }
738     
739     /* put/get bw use opposite streaming/validate nodes */
740     static inline
>>>     CID 188140:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter my_info of type "perf_metrics_t" (size 152 bytes) by value.
741     red_PE_set validation_set(perf_metrics_t my_info, int *nPEs)
742     {
743         if(my_info.cstyle == COMM_PAIRWISE) {
744             if(streaming_node(my_info)) {
745                 *nPEs = my_info.szinitiator;
746                 return FIRST_HALF;

** CID 188139:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/bw_common.h: 64 in print_atomic_header()


________________________________________________________________________________________________________
*** CID 188139:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/bw_common.h: 64 in print_atomic_header()
58     
59     /**************************************************************/
60     /*                   Result Printing and Calc                 */
61     /**************************************************************/
62     
63     static 
>>>     CID 188139:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
64     void print_atomic_header(perf_metrics_t metric_info) {
65         print_header(metric_info);
66         printf("\n\nBandwidth test type:    %10s\n", metric_info.bw_type_str);
67     
68         if (metric_info.cstyle == COMM_INCAST) {
69             printf("Communication style:        INCAST\n");

** CID 188138:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/target_put.h: 85 in target_data_uni_bw()


________________________________________________________________________________________________________
*** CID 188138:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/target_put.h: 85 in target_data_uni_bw()
79             node_to_shadow += my_info.szinitiator;
80         }
81     
82         return partner_nodes;
83     }
84     
>>>     CID 188138:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
85     static inline void target_data_uni_bw(int len, perf_metrics_t metric_info)
86     {
87         double start = 0.0, end = 0.0;
88         int i = 0;
89         unsigned long int j, k;
90         int snode = (metric_info.num_pes != 1)? streaming_node(metric_info) : true;

** CID 188137:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/target_put.h: 28 in get_size_of_side()


________________________________________________________________________________________________________
*** CID 188137:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/target_put.h: 28 in get_size_of_side()
22      * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
23      * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
24      * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
25      * SOFTWARE.
26      */
27     
>>>     CID 188137:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter my_info of type "perf_metrics_t" (size 152 bytes) by value.
28     static inline int get_size_of_side(perf_metrics_t my_info) {
29         if(my_info.my_node < my_info.midpt)
30             return my_info.szinitiator;
31         else
32             return my_info.sztarget;
33     }

** CID 188136:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/target_put.h: 42 in get_num_partners()


________________________________________________________________________________________________________
*** CID 188136:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/target_put.h: 42 in get_num_partners()
36         if(my_info.my_node < my_info.midpt)
37             return my_info.sztarget;
38         else
39             return my_info.szinitiator;
40     }
41     
>>>     CID 188136:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter my_info of type "perf_metrics_t" (size 152 bytes) by value.
42     static inline int get_num_partners(perf_metrics_t my_info, int snode) {
43         int unused_PEs = 0, num_partners = 0;
44         int active_PEs = get_size_of_side(my_info);
45         int other_side = get_size_of_other_side(my_info);
46     
47         if(active_PEs >= other_side) 

** CID 188135:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/int_element_latency.h: 29 in int_p_latency()


________________________________________________________________________________________________________
*** CID 188135:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/int_element_latency.h: 29 in int_p_latency()
23      * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
24      * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
25      * SOFTWARE.
26      */
27     
28     static inline
>>>     CID 188135:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
29     void int_p_latency(perf_metrics_t metric_info)
30     {
31         double start = 0.0;
32         double end = 0.0;
33         unsigned int i = 0;
34         int dest = partner_node(metric_info);

** CID 188134:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/common.h: 781 in print_header()


________________________________________________________________________________________________________
*** CID 188134:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/common.h: 781 in print_header()
775         }
776     
777         *stride = 0; /* back to back PEs */
778     }
779     
780     static
>>>     CID 188134:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter metric_info of type "perf_metrics_t" (size 152 bytes) by value.
781     void print_header(perf_metrics_t metric_info) {
782         printf("\n%20sSandia OpenSHMEM Performance Suite%20s\n", " ", " ");
783         printf("%20s==================================%20s\n", " ", " ");
784         printf("Total Number of PEs:    %10d%6sWindow size:            %10lu\n", 
785                 metric_info.num_pes, " ", metric_info.window_size);
786         printf("Number of source PEs:   %10d%6sMaximum message size:   %10lu\n", 

** CID 188133:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/common.h: 560 in partner_node()


________________________________________________________________________________________________________
*** CID 188133:  Performance inefficiencies  (PASS_BY_VALUE)
/test/performance/shmem_perf_suite/common.h: 560 in partner_node()
554             return 0;
555     }
556     
557     
558     /* Returns partner node; Assumes only one partner */
559     static inline 
>>>     CID 188133:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter my_info of type "perf_metrics_t" (size 152 bytes) by value.
560     int partner_node(perf_metrics_t my_info)
561     {
562         if (my_info.num_pes == 1)
563             return 0;
564     
565         if (my_info.t_type == BW) {
```

**Comment by jdinan** [comment]
GUPS issue to be resolved by #769

**Comment by jdinan** [comment]
Suggest delaying fixes (if any) to performance inefficiencies of passing struct by value to v1.4.x.

**Comment by wrrobin** [comment]
Should be able to fix it soon. Surprised to see that some of them have not been reported earlier as those were not introduced in 1.4.2.

**Comment by jdinan** [comment]
Closed by #770
